File: LowLevelDesign/tic-tac-toe/src/TicTacToe.ts
import { Game } from "./Game";
import { Player } from "./Player";
import { Scoreboard } from "./Scoreboard";
import { Symbol } from "./enum/Symbol";

export class TicTacToe {
    private static instance: TicTacToe;
    private game!: Game;
    private readonly scoreboard: Scoreboard;


    private constructor() {
        // Private constructor to prevent instantiation
        this.scoreboard = new Scoreboard();
    }

    static getInstance(): TicTacToe {
        if (!TicTacToe.instance) {
            TicTacToe.instance = new TicTacToe();
        }
        return TicTacToe.instance;
    }

    public createGame(player1: Player, player2: Player): void {
        this.game = new Game(player1, player2);
        this.game.registerObserver(this.scoreboard);
    }

    public printBoard(): void {
        this.game.board.printBoard();
    }

    public makeMove(row: number, col: number, player: Player): void {
        this.game.makeMove(row, col, player);
    }
}

File: LowLevelDesign/java/tictactoe/src/com/tictac/TicTacToe.java
package com.tictac;

import com.tictac.entities.Game;
import com.tictac.entities.Player;
import com.tictac.observer.GameObserver;
import com.tictac.observer.Scoreboard;

public class TicTacToe {
    private static TicTacToe instance;
    private Game game;
    private GameObserver scoreboard;
    private TicTacToe(){
        this.scoreboard = new Scoreboard();
    }

    public static synchronized TicTacToe getInstance(){
        if(TicTacToe.instance == null){
            TicTacToe.instance = new TicTacToe();
        }
        return TicTacToe.instance;
    }

    public void createGame(Player player1, Player player2) {
        this.game = new Game(player1, player2);
        System.out.printf("Game started between %s (X) and %s (O) %n", player1.getName(), player2.getName());
        game.addObserver(this.scoreboard);
    }

    public void makeMove(int row, int col, Player player) {
        this.game.makeMove(row, col, player);
    }
}
